File: md4.c
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <stdlib.h>
#include <getopt.h>

// Вспомогательные функции F, G, H и циклический сдвиг
// Если x, то y; иначе z. (x & y) | (~x & z)
#define F(x, y, z) ((x & y) | ((~x) & z))
// Мажоритарная функция. (x & y) | (x & z) | (y & z)
#define G(x, y, z) ((x & y) | (x & z) | (y & z))
// Побитовый XOR. x ^ y ^ z
#define H(x, y, z) (x ^ y ^ z)
// Циклический сдвиг влево
#define left_rotate(x, n) ((x << n) | (x >> (32 - n)))
// Преобразование 32 битного регистра в little-endian запись
#define le(x) (((x & 0xFF) << 24) | (((x >> 8) & 0xFF) << 16) | \
    (((x >> 16) & 0xFF) << 8) | ((x >> 24) & 0xFF))


char* md4(const char* message){
    char* hash = malloc(33); // Финальный хэш

    uint32_t msg_len = (uint32_t)strlen(message); // Длина сообщения в байтах

    uint32_t data_len = (msg_len + 64 - (msg_len % 64)); // Длина массива байт
    uint8_t* data = calloc(data_len, sizeof(uint8_t)); // Массив заполненный нулями
    if (!data) { // Не удалось выделить память
        fprintf(stderr, "Sorry, can't allocate memory for you!");
        exit(1);
    }

    memcpy(data, message, msg_len); // Копируем оригинальную строку в начало
    data[msg_len] = 0x80; // Единица и 7 нулей
    uint64_t bit_msg_len = msg_len * 8; // Длина сообщения в битах

    for (int i = 0; i < 8; i++) {
        data[data_len - 8 + i] = (uint8_t)(bit_msg_len >> (i * 8));
    }

    // Регистры ABCD (записаны в Little-endian)
    uint32_t A = 0x67452301; // 01 23 45 67
    uint32_t B = 0xefcdab89; // 89 ab cd ef
    uint32_t C = 0x98badcfe; // fe dc ba 98
    uint32_t D = 0x10325476; // 76 54 32 10

    // Сдвиги раундов
    const uint8_t s1[4] = {3, 7, 11, 19};
    const uint8_t s2[4] = {3, 5, 9, 13};
    const uint8_t s3[4] = {3, 9, 11, 15};

    // Константы раундов
    const uint32_t c2 = 0x5A827999;
    const uint32_t c3 = 0x6ED9EBA1;

    // Порядки слову в раундах
    const uint8_t ord2[16] = {0, 4, 8, 12, 1, 5, 9, 13, 2, 6, 10, 14, 3, 7, 11, 15};
    const uint8_t ord3[16] = {0, 8, 4, 12, 2, 10, 6, 14, 1, 9, 5, 13, 3, 11, 7, 15};

    // Массив под 4 байтовые слова из 16 элеменентов
    uint32_t words[16];

    for (uint32_t ofs=0; ofs < data_len; ofs += 64) {
        // Сохраняем текущие значения регистров
        uint32_t AA = A;
        uint32_t BB = B;
        uint32_t CC = C;
        uint32_t DD = D;

        // Получение текущего 64 байтового блока
        const uint8_t* block = data + ofs;
        
        // Заполнение массива 4 байтовых слов
        for (int i = 0; i < 16; i++) {
            words[i] = (uint32_t)block[i*4] |
            ((uint32_t)block[i*4 + 1] << 8) |
            ((uint32_t)block[i*4 + 2] << 16) |
            ((uint32_t)block[i*4 + 3] << 24);
        }

        uint32_t temp;
        // Раунды 1
        for (uint8_t i=0; i < 16; i++) {
            A = left_rotate(
                (A + F(B, C, D) + words[i]),
                s1[i%4]);
            // Циклически сдвигаем (A,B,C,D) -> (D,A,B,C)
            temp = D;
            D = C;
            C = B;
            B = A;
            A = temp;
            
        }

        // Раунд 2
        for (uint8_t i=0; i < 16; i++) {
            A = left_rotate(
                (A + G(B, C, D) + words[ord2[i]] + c2),
                s2[i%4]);
            // Циклически сдвигаем (A,B,C,D) -> (D,A,B,C)
            temp = D;
            D = C;
            C = B;
            B = A;
            A = temp;
        }

        // Раунд 3
        for (uint8_t i=0; i < 16; i++) {
            A = left_rotate(
                (A + H(B, C, D) + words[ord3[i]] + c3),
                s3[i%4]);
            // Циклически сдвигаем (A,B,C,D) -> (D,A,B,C)
            temp = D;
            D = C;
            C = B;
            B = A;
            A = temp;
        }

        // Обновляем регистры для следующего блока
        A += AA;
        B += BB;
        C += CC;
        D += DD;
    }

    free(data); // Освобождение памяти
    // Формируем хэш как ABCD записанные в little endian
    snprintf(hash, 33, "%08x%08x%08x%08x", le(A), le(B), le(C), le(D));
    return hash; // Хэш
}


int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Usage: %s [-h] message [message ...]\n", argv[0]);
        return 1;
    }
    int opt;
    const char* usage = "usage: md4.py [-h] message [message ...]\n";

    while ((opt = getopt(argc, argv, "h")) != -1) {
        switch (opt) {
            case 'h':
                fprintf(stdout, "%s", usage);
                fprintf(stdout, "\nProgrames create MD4 hash for string(s).\n\npositional arguments:\n" 
                    "  message   The message or messages for which the hash will be calculated.\n\n"
                    "options:\n  -h,   show this help message and exit\n");
                return 0;
            default:
                fprintf(stdout, "%s", usage);
                fprintf(stderr, "%s: error: the following arguments are required: message", argv[0]);
                return 1;
        }
    }

    char* hash;
    for (int i = optind; i < argc; i++) {
        hash = md4(argv[i]);
        fprintf(stdout, "%s\n", hash);
        free(hash);
    }
    return 0;
}

File: md4.py
#!/usr/bin/python3

def md4(message: str|int|float) -> str:
    # Превращаем message в строку
    if not isinstance(message, str):
        message = str(message)
    # Превращаем строку с массив байтов
    message_bytes = message.encode("utf-8")

    length = len(message) # Длина изначального сообщения в байтах

    # Подготовка строки
    data = bytearray(length + 64 - (length % 64)) # Формируем буфер нужного размера
    data[:length] = message_bytes # Записываем в него байты строки
    data[length] = 0x80 # 1000 0000 в Little-endian - единичный бит и нули после него
    data[-8:] = (length*8).to_bytes(8, byteorder="little") # Последние 8 байт - длинa строки

    # Регистры ABCD (записаны в Little-endian)
    A = 0x67452301 # 01 23 45 67
    B = 0xefcdab89 # 89 ab cd ef
    C = 0x98badcfe # fe dc ba 98
    D = 0x10325476 # 76 54 32 10
    MASK = 0xFFFFFFFF # Для приведения к 32 битам (некоторые операции в python выходят за 32 бита)

    # Вспомогательные функции F, G, H и циклический сдвиг
    def F(x, y, z):
        """Если x, то y; иначе z. (x & y) | (~x & z)"""
        return ((x & y) | ((~x) & z)) & MASK
    
    def G(x, y, z):
        """Мажоритарная функция. (x & y) | (x & z) | (y & z)"""
        return ((x & y) | (x & z) | (y & z)) & MASK
    
    def H(x, y, z):
        """Побитовый XOR. x ^ y ^ z"""
        return (x ^ y ^ z) & MASK
    
    def left_rotate(x, n):
        """Циклический сдвиг влево на n бит для 32-битного числа"""
        return ((x << n) | (x >> (32 - n))) & MASK

    # Применение алгоритма к строкам по 512 бит ()
    for ofs in range(0, len(data), 64):
        block = data[ofs: ofs + 64]
        # Список из 16 слов размером 4 байта (32 битные слова)
        words = [int.from_bytes(block[4*i:4*i+4], "little")
                 for i in range(16)]
        # Сохраняем текущие значения регистров
        AA = A
        BB = B
        CC = C
        DD = D

        # Раунд 1 (функция F) - работа с 32-битными словами (4 байта)
        s1 = [3, 7, 11, 19] # Сдвиги раунда
        for i in range(16):
            # A = (A + F(B, C, D) + W[i]) << s1[i]
            A = left_rotate(
                ((A + F(B, C, D) + words[i]) & MASK),
                s1[i%4]);
            A, B, C, D = D, A, B, C # Циклически сдвигаем (A,B,C,D) -> (D,A,B,C)

        # Раунд 2 (функция G) - работа с 32-битными словами (4 байта)
        const_r2 = 0x5A827999 # Константа раунда = 0x5A827999
        order2 = [0, 4, 8, 12, 1, 5, 9, 13, 2, 6, 10, 14, 3, 7, 11, 15] # Порядок слов на раунд
        s2 = [3, 5, 9, 13] # Сдвиги раунда
        for i in range(16):
            # A = (A + G(B, C, D) + W[order2[i]] + C2) << s2[i]
            A = left_rotate(
                ((A + G(B, C, D) + words[order2[i]] + const_r2) & MASK),
                s2[i%4]);
            A, B, C, D = D, A, B, C # Циклически сдвигаем (A,B,C,D) -> (D,A,B,C)

        # Раунд 3
        const_r3 = 0x6ED9EBA1 # Константа раунда = 0x6ED9EBA1
        order3 = [0, 8, 4, 12, 2, 10, 6, 14, 1, 9, 5, 13, 3, 11, 7, 15] # Порядок слов на раунд
        s3 = [3, 9, 11, 15] # Сдвиги раунда
        for i in range(16):
            # A = (A + G(B, C, D) + W[order2[i]] + C3) << s3[i]
            A = left_rotate(
                ((A + H(B, C, D) + words[order3[i]] + const_r3) & MASK),
                s3[i%4]);
            A, B, C, D = D, A, B, C # Циклически сдвигаем (A,B,C,D) -> (D,A,B,C)

        # Обновляем регистры для следующего блока
        A = (A + AA) & MASK
        B = (B + BB) & MASK
        C = (C + CC) & MASK
        D = (D + DD) & MASK

    # Формирование хэша (хэш сформирован как ABCD в little-endian)
    hashed_message = b"".join([
        A.to_bytes(4, byteorder="little"),
        B.to_bytes(4, byteorder="little"),
        C.to_bytes(4, byteorder="little"),
        D.to_bytes(4, byteorder="little"),
    ])
    return hashed_message.hex()


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(
        description="Programes create MD4 hash for string(s).",
        )
    parser.add_argument(
        "message",
        type=str,
        nargs="+",
        help="The message or messages for which the hash will be calculated."
                        )
    args = parser.parse_args()
    for message in args.message:
        print(md4(message))
# 